charts/cluster/templates/_bootstrap.yaml

```diff
@@ -0,0 +1,39 @@
+{{- define "cluster.bootstrap" -}}
+bootstrap:
+{{- if eq .Values.mode "standalone" }}
+  initdb:
+    postInitTemplateSQL:
+    {{- range .Values.cluster.initSQL }}
+      - {{ . }}
+    {{- end -}}
+    {{- if eq .Values.type "postgis" }}
+      - CREATE EXTENSION IF NOT EXISTS postgis;
+      - CREATE EXTENSION IF NOT EXISTS postgis_topology;
+      - CREATE EXTENSION IF NOT EXISTS fuzzystrmatch;
+      - CREATE EXTENSION IF NOT EXISTS postgis_tiger_geocoder;
+    {{- else if eq .Values.type "timescaledb" }}
+      - CREATE EXTENSION IF NOT EXISTS timescaledb;
+  {{- end }}
+{{- else if eq .Values.mode "recovery" }}
+  recovery:
+    {{- with .Values.recovery.pitrTarget.time }}
+    recoveryTarget:
+      targetTime: {{ . }}
+    {{- end }}
+    {{- if eq .Values.recovery.method "backup" }}
+    backup:
+      name: {{ .Values.recovery.backupName }}
+    {{- else if eq .Values.recovery.method "object_store" }}
+    source: objectStoreRecoveryCluster
+    {{- end }}
+
+externalClusters:
+  - name: objectStoreRecoveryCluster
+    barmanObjectStore:
+      serverName: {{ .Values.recovery.serverName }}
+      {{- $d := dict "chartFullname" (include "cluster.fullname" .) "scope" .Values.recovery "secretSuffix" "-recovery" -}}
+      {{- include "cluster.barmanObjectStoreConfig" $d | nindent 4 }}
+{{-  else }}
+  {{ fail "Invalid cluster mode!" }}
```

Comment:
I see you are using fails, what about defining a [schema](https://helm.sh/docs/topics/charts/#schema-files) instead? you could define enums for these fields and avoid having to deal with wrong fields, unfortunately the tool we were using to automatically generate schemas was archived, but I forked it [here](https://github.com/phisco/helm-schema-gen) to make it run on ARM based macbooks too at least, but it doesn't support enums out of the box, so it's mainly going to be useful as a starting point, if you want below you can find the output of `helm schema-gen charts/cluster/values.yaml` with the current values.yaml:
```json
{
    "$schema": "http://json-schema.org/schema#",
    "type": "object",
    "properties": {
        "backups": {
            "type": "object",
            "properties": {
                "azure": {
                    "type": "object",
                    "properties": {
                        "connectionString": {
                            "type": "string"
                        },
                        "containerName": {
                            "type": "string"
                        },
                        "inheritFromAzureAD": {
                            "type": "boolean"
                        },
                        "path": {
                            "type": "string"
                        },
                        "serviceName": {
                            "type": "string"
                        },
                        "storageAccount": {
                            "type": "string"
                        },
                        "storageKey": {
                            "type": "string"
                        },
                        "storageSasToken": {
                            "type": "string"
                        }
                    }
                },
                "destinationPath": {
                    "type": "string"
                },
                "enabled": {
                    "type": "boolean"
                },
                "endpointURL": {
                    "type": "string"
                },
                "google": {
                    "type": "object",
                    "properties": {
                        "applicationCredentials": {
                            "type": "string"
                        },
                        "bucket": {
                            "type": "string"
                        },
                        "gkeEnvironment": {
                            "type": "boolean"
                        },
                        "path": {
                            "type": "string"
                        }
                    }
                },
                "provider": {
                    "type": "string"
                },
                "retentionPolicy": {
                    "type": "string"
                },
                "s3": {
                    "type": "object",
                    "properties": {
                        "accessKey": {
                            "type": "string"
                        },
                        "bucket": {
                            "type": "string"
                        },
                        "path": {
                            "type": "string"
                        },
                        "region": {
                            "type": "string"
                        },
                        "secretKey": {
                            "type": "string"
                        }
                    }
                },
                "scheduledBackups": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "backupOwnerReference": {
                                "type": "string"
                            },
                            "name": {
                                "type": "string"
                            },
                            "schedule": {
                                "type": "string"
                            }
                        }
                    }
                }
            }
        },
        "cluster": {
            "type": "object",
            "properties": {
                "additionalLabels": {
                    "type": "object"
                },
                "affinity": {
                    "type": "object",
                    "properties": {
                        "topologyKey": {
                            "type": "string"
                        }
                    }
                },
                "annotations": {
                    "type": "object"
                },
                "certificates": {
                    "type": "null"
                },
                "enableSuperuserAccess": {
                    "type": "boolean"
                },
                "imageName": {
                    "type": "string"
                },
                "imagePullPolicy": {
                    "type": "string"
                },
                "imagePullSecrets": {
                    "type": "array"
                },
                "initSQL": {
                    "type": "array"
                },
                "instances": {
                    "type": "integer"
                },
                "logLevel": {
                    "type": "string"
                },
                "monitoring": {
                    "type": "object",
                    "properties": {
                        "customQueries": {
                            "type": "array"
                        },
                        "enablePodMonitor": {
                            "type": "boolean"
                        },
                        "grafanaDashboard": {
                            "type": "object",
                            "properties": {
                                "enable": {
                                    "type": "boolean"
                                }
                            }
                        }
                    }
                },
                "postgresql": {
                    "type": "null"
                },
                "primaryUpdateMethod": {
                    "type": "string"
                },
                "primaryUpdateStrategy": {
                    "type": "string"
                },
                "priorityClassName": {
                    "type": "string"
                },
                "resources": {
                    "type": "null"
                },
                "storage": {
                    "type": "object",
                    "properties": {
                        "size": {
                            "type": "string"
                        },
                        "storageClass": {
                            "type": "string"
                        }
                    }
                },
                "superuserSecret": {
                    "type": "string"
                }
            }
        },
        "fullnameOverride": {
            "type": "string"
        },
        "mode": {
            "type": "string"
        },
        "nameOverride": {
            "type": "string"
        },
        "pooler": {
            "type": "object",
            "properties": {
                "enabled": {
                    "type": "boolean"
                },
                "instances": {
                    "type": "integer"
                },
                "parameters": {
                    "type": "object",
                    "properties": {
                        "default_pool_size": {
                            "type": "string"
                        },
                        "max_client_conn": {
                            "type": "string"
                        }
                    }
                },
                "poolMode": {
                    "type": "string"
                }
            }
        },
        "recovery": {
            "type": "object",
            "properties": {
                "azure": {
                    "type": "object",
                    "properties": {
                        "connectionString": {
                            "type": "string"
                        },
                        "containerName": {
                            "type": "string"
                        },
                        "inheritFromAzureAD": {
                            "type": "boolean"
                        },
                        "path": {
                            "type": "string"
                        },
                        "serviceName": {
                            "type": "string"
                        },
                        "storageAccount": {
                            "type": "string"
                        },
                        "storageKey": {
                            "type": "string"
                        },
                        "storageSasToken": {
                            "type": "string"
                        }
                    }
                },
                "backupName": {
                    "type": "string"
                },
                "clusterName": {
                    "type": "string"
                },
                "destinationPath": {
                    "type": "string"
                },
                "endpointURL": {
                    "type": "string"
                },
                "google": {
                    "type": "object",
                    "properties": {
                        "applicationCredentials": {
                            "type": "string"
                        },
                        "bucket": {
                            "type": "string"
                        },
                        "gkeEnvironment": {
                            "type": "boolean"
                        },
                        "path": {
                            "type": "string"
                        }
                    }
                },
                "method": {
                    "type": "string"
                },
                "pitrTarget": {
                    "type": "object",
                    "properties": {
                        "time": {
                            "type": "string"
                        }
                    }
                },
                "provider": {
                    "type": "string"
                },
                "s3": {
                    "type": "object",
                    "properties": {
                        "accessKey": {
                            "type": "string"
                        },
                        "bucket": {
                            "type": "string"
                        },
                        "path": {
                            "type": "string"
                        },
                        "region": {
                            "type": "string"
                        },
                        "secretKey": {
                            "type": "string"
                        }
                    }
                }
            }
        },
        "type": {
            "type": "string"
        }
    }
}
```

Reply:
I am not sure about he schema proposal. There are a lot of different validation configurations. I'd rather handle them in an:

```
if option1
else if option2
else
  fail
```

approach.

The problem is with nested checks configurations, for example certain fields need to be set only if backups are enabled. I think this way in general we can produce more helpful messages to guide the user.

Also it will be easier to maintain. Could we stick with this solution, at least until the chart has reached a stable version and we know we won't change configuration structure?

Reply:
Yes, you'd still need to specify those conditional checks via the chart, I was mostly suggesting it for the enums and plain required fields, to avoid having the else branches dealing with invalid values. But given that it would be mostly a manual thing I agree we can postpone that to when the values will be more stable 👍